src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,795 @@
+        Kernel<BinaryOpBackwardUseNone<LOP, Req>, xpu>::Launch(s, size, lgrad_dptr, ograd_dptr);
+      });
+    }
+    if (std::is_same<ROP, mshadow_op::identity>::value && req[1] == kWriteInplace) {
+      CHECK_EQ(ograd_dptr, outputs[1].dptr<DType>());
+    } else if (req[1] != kNullOp) {
+      DType *rgrad_dptr = outputs[1].dptr<DType>();
+      MXNET_ASSIGN_REQ_SWITCH(req[1], Req, {
+        Kernel<BinaryOpBackwardUseNone<ROP, Req>, xpu>::Launch(s, size, rgrad_dptr, ograd_dptr);
+      });
+    }
+  }
+
+  template<typename xpu, typename LOP, typename ROP, typename DType>
+  static void BinaryBackwardUseIn_(const nnvm::NodeAttrs &attrs,
+                                   const OpContext &ctx,
+                                   const std::vector<TBlob> &inputs,
+                                   const std::vector<OpReqType> &req,
+                                   const std::vector<TBlob> &outputs) {
+    DCHECK_EQ(outputs.size(), 2U);
+    DCHECK_EQ(inputs.size(), 3U);
+    mxnet_op::Stream<xpu> *s = ctx.get_stream<xpu>();
+    const DType *ograd_dptr = inputs[0].dptr<DType>();
+    const DType *lhs_dptr = inputs[1].dptr<DType>();
+    const DType *rhs_dptr = inputs[2].dptr<DType>();
+    MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+      const int size = static_cast<int>(
+        (outputs[0].Size() + mxnet_op::DataType<DType>::kLanes - 1)
+        / mxnet_op::DataType<DType>::kLanes);
+      DType * lgrad_dptr = outputs[0].dptr<DType>();
+      mxnet_op::Kernel<BinaryOpBackwardUseIn<LOP, Req>, xpu>::Launch(
+        s, size, lgrad_dptr, ograd_dptr, lhs_dptr, rhs_dptr);});
+    MXNET_ASSIGN_REQ_SWITCH(req[1], Req, {
+      const int size = static_cast<int>(
+        (outputs[1].Size() + mxnet_op::DataType<DType>::kLanes - 1)
+        / mxnet_op::DataType<DType>::kLanes);
+      DType * rgrad_dptr = outputs[1].dptr<DType>();
+      mxnet_op::Kernel<BinaryOpBackwardUseIn<ROP, Req>, xpu>::Launch(
+        s, size, rgrad_dptr, ograd_dptr, lhs_dptr, rhs_dptr);});
+  }
+
+  template<
+    typename xpu,
+    typename LOP,
+    typename ROP,
+    typename DType,
+    bool in0_ok_dense = false,
+    bool in1_ok_dense = false,
+    bool in2_ok_dense = false,
+    typename BackupCompute>
+  static inline void BinaryBackwardUseInEx_(const nnvm::NodeAttrs &attrs,
+                                           const OpContext &ctx,
+                                           const std::vector<NDArray> &inputs,
+                                           const std::vector<OpReqType> &req,
+                                           const std::vector<NDArray> &outputs,
+                                           BackupCompute backup_compute) {
+    CHECK_EQ(inputs.size(), 3U);  // output grad,
+    CHECK_EQ(outputs.size(), 2U);  // lhs input grad, rhs input grad
+    if (req[0] != kNullOp) {
+      // If any input is dense, fallback to FCompute
+      if (!common::ContainsDefaultStorage(inputs)) {
+        mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
```

Comment:
It looks like this is implemented for rsp only. could you add CHECK_EQ in case there's other spares stype here?